File: 5123.js
/**
 * @param {string} characters
 * @param {number} combinationLength
 */
var CombinationIterator = function(characters, combinationLength) {
    this.characters = characters;
    this.strLen = characters.length;
    this.combinationLength = combinationLength;
    this.indexArr = new Array(combinationLength);
    this.nextStr = characters.substring(0,combinationLength);
    for(let i = 0;i<combinationLength;i++){
        this.indexArr[i] = i;
    }
};

/**
 * @return {string}
 */
CombinationIterator.prototype.next = function() {
    const ret = this.nextStr;
    let find = false;
    for(let j=this.combinationLength-1,k=0;j>=0;j--,k++){
        if(this.indexArr[j] + k <this.strLen-1){
            this.indexArr[j] = this.indexArr[j] + 1;
            for(let i = j+1;i<this.combinationLength;i++){
                this.indexArr[i] = this.indexArr[i-1]+1;
            }
            find = true;
            break;
        }
    }
    if(!find){
        this.nextStr = null;
    } else {
        let strArray = "";
        for(let i=0;i<this.combinationLength;i++){
            strArray += this.characters[this.indexArr[i]];
        }
        this.nextStr = strArray;
    }
    return ret;
};

/**
 * @return {boolean}
 */
CombinationIterator.prototype.hasNext = function() {
    return this.nextStr != null;
};

/** 
 * Your CombinationIterator object will be instantiated and called as such:
 * var obj = new CombinationIterator(characters, combinationLength)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */

var obj = new CombinationIterator("abcdefg",2)
//var param_1 = obj.next()
//var param_2 = obj.hasNext()
while(obj.hasNext()){
    console.log(obj.next());
}

File: 5123.c



typedef struct {
    char characters[15];
    int strLen;
    int combinationLength;
    int indexArr[15];
    char nextStr[15];
    int hasNext;
} CombinationIterator;


CombinationIterator* combinationIteratorCreate(char * characters, int combinationLength) {
    CombinationIterator * ret = (CombinationIterator *)malloc(sizeof(CombinationIterator));
    memcpy(ret->characters,characters);
    ret->combinationLength = combinationLength;
    ret->strLen = strlen(characters);
    for(int i = 0;i<combinationLength;i++){
        ret->indexArr[i] = i;
    }
    ret->hasNext = 1;
    return ret;
}

char * combinationIteratorNext(CombinationIterator* obj) {
    char ret[15];
    strcpy(ret,obj->nextStr);
    int find = 0;
    for(int j=obj->combinationLength-1;j>=0;j--){
            if(obj->indexArr[j]<obj->strLen-(obj->combinationLength-j)){
                obj->indexArr[j] = obj->indexArr[j] + 1;
                for(int i = j+1;i<obj->combinationLength;i++){
                    obj->indexArr[i] = obj->indexArr[i-1]+1;
                }
                obj->indexArr[j] = obj->indexArr[j]+1;
                find = 1;
            }
    }
    if(!find){
        obj->hasNext = 0;
    } else {
        for(int i=0;i<obj->combinationLength;i++){
            obj->nextStr[i] = obj->characters[obj->indexArr[i]];
        }
    }
    return ret;
}

bool combinationIteratorHasNext(CombinationIterator* obj) {
  return (bool)obj->hasNext;
}

void combinationIteratorFree(CombinationIterator* obj) {
    free(obj);
    obj = 0;
}

/**
 * Your CombinationIterator struct will be instantiated and called as such:
 * CombinationIterator* obj = combinationIteratorCreate(characters, combinationLength);
 * char * param_1 = combinationIteratorNext(obj);
 
 * bool param_2 = combinationIteratorHasNext(obj);
 
 * combinationIteratorFree(obj);
*/